### PR TITLE
M3.5: Post-M3 Cleanup

### LINKED ISSUE #120: Task: M3 Basic docs update
# Task: M3 Basic docs update

## Summary

Update README with basic job execution documentation.

## Parent / grouping

- Milestone / Parent issue: #85 (M3: Job Execution MVP)

## Requirements

- [ ] Add "Running Jobs" section to README
- [ ] Document `bosun job run` usage
- [ ] Document `--dry-run` flag
- [ ] Add example compose file with job labels
- [ ] Verify `make docs` still works

## Non-goals

- Comprehensive user guide (M6)
- Example worker images (M6)

## Dependencies (Linked Issues)

- Blocked by:
  - `bosun job run` CLI command
- Blocks:
  - (none)

## Acceptance criteria

- [ ] README has job execution section
- [ ] Examples are accurate and tested
- [ ] No broken links

## Validation / test plan

- How to verify: Follow README instructions on fresh setup
- Edge cases: N/A

## Notes

- Keep brief; comprehensive docs are M6

### LINKED ISSUE #133: repeated error message in case of unvalid arg
```
❯ ./bin/bosun completition
Error: unknown command "completition" for "bosun"

Did you mean this?
        completion

Run 'bosun --help' for usage.
2026/01/03 19:18:02 unknown command "completition" for "bosun"

Did you mean this?
        completion
```

### LINKED ISSUE #139: Fix invalid labels in test compose file (bosun.other, bosun.network)
# Task: Fix invalid labels in test compose file

## Summary

The test compose file `internal/testutil/compose/joblabels-compose.yaml` contains invalid labels (`bosun.other`, `bosun.network`) that cause validation failures. These are test artifacts, not design features.

## Parent / grouping

- Milestone / Parent issue: #138 (M3.5: Post-M3 Cleanup)
- Related: #136 (validation fails with test compose)

## Requirements

- [ ] Remove `bosun.other: "value"` from `joblabels-compose.yaml:46`
- [ ] Change `bosun.network: "true"` to valid `bosun.network.priority: "50"` at line 68
- [ ] Verify `bosun config validate` passes on all test compose files
- [ ] Update any tests that expect these invalid labels

## Non-goals

- Adding new label namespaces (e.g., `bosun.other.*`)
- Changing validation to allow arbitrary labels

## Evidence

```yaml
# joblabels-compose.yaml
no-job-labels:
  labels:
    bosun.other: "value"    # ← Invalid - not in schema

networks:
  backend:
    labels:
      bosun.network: "true"   # ← Invalid - should be bosun.network.priority
```

## Acceptance criteria

- [ ] `bosun config validate` passes on `joblabels-compose.yaml`
- [ ] All integration tests pass
- [ ] No unknown key errors for test compose files

## Validation / test plan

- How to verify: `bin/bosun config validate --project joblabels` returns no errors
- Edge cases: N/A

## Notes

- This directly fixes #136
- The valid label is `bosun.network.priority` (defined in schema)

### LINKED ISSUE #140: Rename "backup" terminology to generic "job" terminology
# Task: Rename "backup" terminology to generic "job" terminology

## Summary

Bosun is a **generic job orchestrator**, NOT a backup tool. However, the codebase is littered with "backup" terminology that implies a specific use case. This creates confusion and limits perceived applicability.

## Parent / grouping

- Milestone / Parent issue: #138 (M3.5: Post-M3 Cleanup)

## Requirements

Production code changes (tests are OK to keep "backup" as example use case):

- [ ] `internal/config/schema/config_v1.go`: Rename `BackupEnabled` → `Enabled` or `JobEnabled`
- [ ] `internal/config/schema/config_v1.go`: Update label `bosun.volume.backupEnabled` → `bosun.volume.enabled`
- [ ] `internal/config/schema/job_labels.go`: Update comments (8 occurrences of "backup")
- [ ] `internal/domain/jobs/run.go`: Fix package doc "backup job execution" → "job execution"
- [ ] `internal/adapters/docker/worker/doc.go`: Fix "backup worker" → "worker" references
- [ ] `internal/cmd/plan_list.go`: Fix "List discovered backup jobs" → "List discovered jobs"
- [ ] `internal/cmd/plan_show.go`: Fix "backup job" → "job"
- [ ] `internal/cmd/job_run.go`: Fix "backup job" references
- [ ] `internal/ports/doc.go`: Fix "backup job definitions" → "job definitions"
- [ ] `docs/config.md`: Update documentation

## Non-goals

- Renaming test strings like `"daily-backup"`, `"backup-job"` — these are valid use case examples

## Acceptance criteria

- [ ] No production code contains "backup" except in example/doc contexts
- [ ] All tests pass
- [ ] `bosun.volume.backupEnabled` label still works (backward compat alias) OR migration documented

## Validation / test plan

- How to verify: `grep -r "backup" internal/ --include="*.go" | grep -v "_test.go"` returns only acceptable occurrences
- Edge cases: Existing compose files with `bosun.volume.backupEnabled` — need migration path

## Notes

- This is a **breaking change** if we remove `backupEnabled` label without alias
- Consider: add alias in loader to accept both old and new label for one release cycle

### LINKED ISSUE #142: ExecutionPlan should be authoritative (executor ignores plan steps)
# Task: Make ExecutionPlan authoritative

## Summary

The current design creates an `ExecutionPlan` but the executor ignores it and hardcodes the execution flow. The plan should drive execution, not just be a preview.

## Parent / grouping

- Milestone / Parent issue: #138 (M3.5: Post-M3 Cleanup)
- Related: #135 (missing restart step in plan display)

## Current behavior (broken)

```go
// executor.go - ExecuteJob()
plan, _ := e.planner.Plan(ctx, job)  // Plan is created...

// ...but then executor hardcodes the flow:
e.compose.StopStack(...)   // Hardcoded step 1
e.worker.Run(...)          // Hardcoded step 2
defer e.compose.StartStack(...) // Hardcoded step 3 (in defer)
```

The plan is only used for logging, NOT for driving execution.

## Expected behavior

```go
// Executor should interpret plan steps
plan, _ := e.planner.Plan(ctx, job)
for _, step := range plan.Steps {
    result := e.executeStep(ctx, step)
    // record result
}
```

## Requirements

- [ ] Planner adds `StartContainers` step to plan (currently omitted)
- [ ] Executor iterates plan steps instead of hardcoding flow
- [ ] Each step type has a handler in executor
- [ ] Step results are recorded per-step
- [ ] DryRun and Execute use the same plan (consistency)

## Non-goals

- Per-step retry policies (defer to M4)
- Parallel step execution

## Acceptance criteria

- [ ] `bosun plan show <job>` displays start/restart step
- [ ] Executor executes exactly what the plan shows
- [ ] DryRun plan matches actual execution flow
- [ ] All tests pass

## Validation / test plan

- How to verify: `bosun plan show job --format json` includes `start_containers` step
- Edge cases: Plan with zero steps, plan with only worker step

## Notes

- This is a design fix, not just #135 (which is just the display issue)
- Consider: `PlanStep` may need policy metadata (timeouts, retries) in future

### LINKED ISSUE #143: Fix JobExecutor interface mismatch (Execute vs ExecuteJob)
# Task: Fix JobExecutor interface mismatch

## Summary

The `ports.JobExecutor` interface defines `Execute(ctx, jobName string, ...)` but the implementation provides `ExecuteJob(ctx, job jobs.Job, ...)`. The interface and implementation don't match.

## Parent / grouping

- Milestone / Parent issue: #138 (M3.5: Post-M3 Cleanup)

## Current state

```go
// ports/executor.go - Interface expects job lookup by name
type JobExecutor interface {
    Execute(ctx context.Context, jobName string, opts ExecuteOptions) (ExecutionResult, error)
    DryRun(ctx context.Context, jobName string) (ExecutionPlan, error)
}

// app/executor/executor.go - Implementation accepts Job directly
func (e *Executor) ExecuteJob(ctx context.Context, job jobs.Job, ...) { ... }
func (e *Executor) Execute(...) { return error("not implemented") }  // Stub!
```

The `Execute` method returns "not implemented" and constructor accepts unused `discoverer` parameter.

## Recommended fix

Change interface to accept `jobs.Job` (Option B - more modular):

```go
type JobExecutor interface {
    Execute(ctx context.Context, job jobs.Job, opts ExecuteOptions) (ExecutionResult, error)
    DryRun(ctx context.Context, job jobs.Job) (ExecutionPlan, error)
}
```

**Rationale:** Single responsibility - executor executes, discovery is separate.

## Requirements

- [ ] Update `ports.JobExecutor` interface to accept `jobs.Job`
- [ ] Rename `ExecuteJob` → `Execute` in executor
- [ ] Rename `DryRunJob` → `DryRun` in executor
- [ ] Remove unused `discoverer` parameter from `executor.New()`
- [ ] Update all callers

## Non-goals

- Adding job discovery to executor (keep separate)

## Acceptance criteria

- [ ] Interface matches implementation
- [ ] No unused constructor parameters
- [ ] No "not implemented" stub methods
- [ ] All tests pass

## Validation / test plan

- How to verify: `go build ./...` succeeds, interface is satisfied
- Edge cases: N/A

## Notes

- Alternative (Option A): Implement `Execute(jobName)` using injected discoverer
- Team chose Option B for modularity

### PR DESCRIPTION
**Pull request by simone-viozzi**
This PR addresses code smells and design inconsistencies identified after the M3 job execution MVP delivery. The changes improve validation UX, fix terminology confusion, and align implementation with architectural principles.

## Context

After completing the job execution MVP (M3), a systematic smell review identified several issues: the validation system rejected unknown labels too strictly, "backup" terminology throughout the codebase implied Bosun was backup-specific rather than a generic job orchestrator, and the execution plan was cosmetic rather than authoritative.

### Lenient validation with strict mode option
Unknown `bosun.*` label keys now generate warnings instead of errors by default, following Docker/Kubernetes precedent for extensibility. A new `--strict` flag restores hard failures when needed. The validation error type was restructured to distinguish errors from warnings.

Fixes #139, #136, #134

### Generic job terminology
Removed "backup" references from production code (schema, domain, adapters, CLI descriptions). Config field `BackupEnabled` renamed to `Enabled`, label `bosun.volume.backupEnabled` renamed to `bosun.volume.enabled`. Test fixtures and example names like "daily-backup" remain as valid use-case examples.

Fixes #140

### Plan-driven execution
The planner now generates a complete execution plan including the `start_containers` step (previously omitted). The executor interprets plan steps instead of hardcoding the sequence, ensuring DryRun previews match actual execution.

Fixes #142, #135

### Clean interface contracts
The `JobExecutor` interface now accepts `jobs.Job` directly instead of job names, following single-responsibility principle. Removed unused `discoverer` constructor parameter and stub methods. Renamed `ExecuteJob/DryRunJob` to `Execute/DryRun` for consistency.

Fixes #143

### CLI error deduplication
Cobra's automatic error printing disabled to prevent duplicate messages. Errors now print once without log timestamps.

Fixes #133

### Field naming consistency
Renamed `PlanStep.UseComposeStop` to `UseCompose` since the field applies to both stop and start operations.

### Testing improvements
- Added test worker Docker image for integration testing
- Updated integration tests for lenient validation mode
- Fixed invalid test labels in compose files

Fixes #120

## Breaking Changes

**Config schema:**
- `VolumeConfig.BackupEnabled` → `VolumeConfig.Enabled`
- Label `bosun.volume.backupEnabled` → `bosun.volume.enabled`

**JSON output:**
- Plan step field `useComposeStop` → `useCompose`

**Internal API (pre-1.0):**
- `JobExecutor.ExecuteJob()` → `JobExecutor.Execute()`
- `JobExecutor.DryRunJob()` → `JobExecutor.DryRun()`
- Methods now accept `jobs.Job` parameter instead of job name string

**Validation behavior:**
- Unknown `bosun.*` keys are warnings by default (use `--strict` for errors)

## Documentation

Project constitution updated to v1.2.0, codifying Plan-Driven Execution principle and lenient validation policy. Serena memories updated to reflect new package APIs.

---

Closes #133, #135, #139, #140, #142, #143  
Fixes #120, #134, #136

### CODE REVIEW

File: internal/app/planner/planner.go [hunk 1]
```diff
@@ -80,9 +80,28 @@ func (p *Planner) Plan(ctx context.Context, job jobs.Job) (jobs.ExecutionPlan, e
...
+			useComposeStart = true
+			composeProject = job.TargetStacks[0]
+		}
+
+		startStep := jobs.PlanStep{
+			Type:           jobs.StepTypeStartContainers,
+			Description:    generateStartDescription(containerNames, useComposeStart, composeProject),
+			ContainerIDs:   targetContainers,
+			ContainerNames: containerNames,
+			UseComposeStop: useComposeStart, // reuses field for compose start
```
[the remark this answers is not in the document]
Comment (simone-viozzi) [reply internal/app/planner/planner.go]: should be solved in [31850c9](https://github.com/simone-viozzi/bosun/pull/151/commits/31850c9172f62bf5eca62d23811e0fd4d9edcc36)

File: internal/app/planner/planner.go [hunk 2]
```diff
@@ -80,9 +80,28 @@ func (p *Planner) Plan(ctx context.Context, job jobs.Job) (jobs.ExecutionPlan, e
...
 	steps = append(steps, runWorkerStep)
 
-	// TODO(#142): Add StepTypeStartContainers step here, make ExecutionPlan authoritative.
-	// Executor should interpret plan.Steps, not use hardcoded sequence.
-	// See smells #23-24 in wip_smell_milestone3
+	// Step 3: Start containers (if any were stopped)
+	if len(targetContainers) > 0 {
+		containerNames := extractContainerNames(targetContainers)
+
+		// Use compose start if we used compose stop
```
[the remark this answers is not in the document]
Comment (simone-viozzi) [reply internal/app/planner/planner.go]: should be fixed in [7b6b9da](https://github.com/simone-viozzi/bosun/pull/151/commits/7b6b9da7db399e0560733ea20c15e2b61d76128d)

File: internal/cmd/job_run.go
[the remark this answers is not in the document]
Comment (simone-viozzi) [reply internal/cmd/job_run.go]: should be fixed in [7b6b9da](https://github.com/simone-viozzi/bosun/pull/151/commits/7b6b9da7db399e0560733ea20c15e2b61d76128d)
